stdlib/public/core/StringStorage.swift

```diff
@@ -23,6 +23,12 @@ class _SwiftRawStringStorage : _SwiftNativeNSString {
   public // @testable
   final var count: Int
 
+  // TODO: Is there a better way to derive this?
+  @nonobjc
+  @_versioned
+  internal static
+  let _storageOffset: Int = MemoryLayout<Int>.stride * 4
```

Comment:
Unfortunately I think this is the best we have today. We don't have an equivalent of `offsetof` in Swift, and you were going to access it directly anyway. It's not right in 32-bit code, though, is it?

Reply:
Argh, github isn't highlighting what code your previous comment is applying to. For the `isUnmangedOrNative`sanity check changes, I think they're all correct. Basically, "unmanaged" is less of a distinct form, and more of an immortal variant on native. Distinguishing between them is only interesting for append-like operations that may want to grow. Growth cannot kill an immortal :-)

When we want to read from a contiguous, non-Cocoa non-Small string, we extract it into an `_UnmanagedString` efficiently by adding the bias to the first word's payload and using the second word's low bits as the count. Mortality is irrelevant to reads :-)

Ideally, I'd revise all the naming and comments here to accommodate this. Any ideas?

* "normal" form, which would be comically overloaded.
* "contiguous" form, which may be unclear as most lazily-bridged NSStrings are contiguous.
* "(im)mortal" form, which can only be spelled that way in comments.

---

Regarding `.string`, right you are!

---

Regarding the offset, I think you're right. I was hoping that counting words was portable, but I might just have to `#if` it.

Reply:
Actually, HeapObject's inline count is defined in terms of words, so this should be the correct approach. 32-bit is failing currently because StringObject's enum doesn't lay native and unmanaged the same. That's an easy fix.

Reply:
You're still going to need an extra word for flags on 32-bit, right?

Reply:
We need the extra word in _StringGuts, but not inside the storage object.

What if we would call the biased start + length form “canonical”? Or perhaps “regular” would work even better.

Reply:
It seems like it's worth saving a word inside the storage object, but I guess you want to keep things aligned, and I suppose you can always put stuff *after* the string data if you really need to.

I won't comment on the naming and the rest of it. I just don't like this intermediate state where some places say "unmanaged-or-native" and others just say "unmanaged". Either you've collapsed them or you haven't.